keps/sig-api-machinery/3945-crd-conversion-with-cel/README.md

```diff
@@ -0,0 +1,912 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+This KEP is part of a broader effort to leverage CEL as an embedded expression language in Kubernetes to handle tasks that currently require admission control via a webhook.
+
+CRD validation rules has successful brought CEL into CRD validation which provides an alternative to validating admission webhooks.
+This KEP focuses on CRD conversion which would leverage capabilities and best practices learned from CRD validation rules (introduced in v1.23).
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+This KEP is primarily motivated by the operational difficulties encountered using webhooks to perform CRD version conversion. 
+And with leveraging the capability of CEL, it will improve the simplicity of CRD development as well as making CRD more self-contained.
+
+TODO: add info about current conversion webhook
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Make CRDs more self-contained and declarative
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. 
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+The [Common Expression Language (CEL)](https://github.com/google/cel-go) would be used for version conversion which
+currently can only be done through conversion webhooks.
+
+Current the conversion is done through the conversion strategy in CRD which supports 'none' and 'webhook'. 
+This KEP proposes to add `x-kubernetes-conversion` to CRD structural schemas to allow CEL conversion.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          type: object
+          properties:
+            demoField:
+              type: integer
+              x-kubernetes-conversion:
+                - from: v1 # from is the source version.
+                  to: v2 # to is the target version.
```

Comment:
For native types, we define the `internal` type to be the `hub` type for conversion.  So if you convert from v1alpha1 to v1, the process is `v1alpha1->hub->v1`.

We don't have hub types for CRDs, but thinking about this made me wonder.  If you have three types, say `v1alpha1`, `v1beta1` and `v1`.  What conversions are you required to define?  All 6 possible transitions? Or just enough so that conversions can be chained to get to the version that you need?   If you say, defined:

`from: v1alpha1, to: v1`, `from: v1, to: v1alpha1` 
`from: v1beta1, to: v1`,  `from: v1, to: v1beta1` 

Would the system be smart enough to convert from `v1beta1` to `v1alpha1` for you?  

(and if there are multiple possible conversion paths, which one is used?)